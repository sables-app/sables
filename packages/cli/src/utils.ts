import type { SablesBuildOptions } from "./types.js";

/** @public */
export const serverTargets = {
  /**
   * Server build target for the Node.js runtime
   *
   * @public
   */
  NODE: "node",
  /**
   * Server build target for the Cloudflare Workers runtime
   *
   * @public
   */
  WORKER: "worker",
} as const;

/**
 * A utility to retrieve boolean values from the Node.js env.
 * Intended to be used within Sables configuration files.
 *
 * @public
 */
export function envFlag(name: string) {
  return !!JSON.parse(process.env[name] || "false");
}

/**
 * A utility to assist in writing configuration for Sables.
 *
 * @public
 */
export function defineConfig(config: SablesBuildOptions): SablesBuildOptions {
  return config;
}

defineConfig.serverTargets = serverTargets;
