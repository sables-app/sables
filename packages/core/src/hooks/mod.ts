export * from "./Action.js";
export * from "./Observable.js";
export * from "./Selector.js";
