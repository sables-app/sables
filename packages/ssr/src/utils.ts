import { SSR_ATTRIBUTE, ssrAttrValues } from "@sables/framework";

export function shouldHydrate() {
  return (
    typeof document == "object" &&
    !!document.querySelector(
      `[${SSR_ATTRIBUTE}="${ssrAttrValues.SHOULD_HYDRATE}"]`
    )
  );
}
