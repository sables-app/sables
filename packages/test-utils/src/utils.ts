import { VITest } from "./types.js";

type LogKey = "info" | "log" | "warn" | "error";

export function silenceLogs(vitest: VITest, key: LogKey) {
  const { afterAll, beforeAll, vi } = vitest;
  let originalMethod: typeof console[LogKey];

  beforeAll(() => {
    originalMethod = console[key];
    console[key] = vi.fn();
  });

  afterAll(() => {
    console[key] = originalMethod;
  });
}
