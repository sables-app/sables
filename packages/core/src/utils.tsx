import { createMutableRef, MutableReferenceObjectValue } from "@sables/utils";

import type * as ReduxToolkit from "@reduxjs/toolkit";
import type { ComponentProps, ElementType, FunctionComponent } from "react";
import type * as Redux from "redux";

import {
  LIFECYCLE_REF_MESSAGE,
  ROUTES_COLLECTION_REF_MESSAGE,
  SYMBOL_EFFECT_API_LIFECYCLE,
  SYMBOL_EFFECT_API_ROUTES,
  SYMBOL_LAZY_META,
  SYMBOL_LAZY_SELECTOR,
} from "./constants.js";
import { ActionSubject } from "./main.js";
import type {
  AnyLazySelector,
  DefaultEffectAPI,
  LazyMeta,
  ObjectWithLazyMeta,
} from "./types.js";

/** @internal */
export function hasLazyMeta<T extends object>(
  value: T
): value is T & ObjectWithLazyMeta {
  return Object.hasOwn(value, SYMBOL_LAZY_META);
}

/** @internal */
export function createLazyMeta(): LazyMeta {
  return {
    observers: new Set(),
    slices: new Set(),
  };
}

/** @internal */
export function isEnhancedSelector(value: unknown): value is AnyLazySelector {
  return (
    typeof value == "function" && Object.hasOwn(value, SYMBOL_LAZY_SELECTOR)
  );
}

/** @internal */
export function getSlicesFromSelector(
  selector: ReduxToolkit.Selector
): ReduxToolkit.Slice[] {
  return isEnhancedSelector(selector)
    ? [...selector[SYMBOL_LAZY_META].slices]
    : [];
}

type LifecycleRef<
  StoreState extends Record<string, unknown> = Record<string, unknown>
> = DefaultEffectAPI<StoreState>[typeof SYMBOL_EFFECT_API_LIFECYCLE];

type RoutesCollectionRef<
  StoreState extends Record<string, unknown> = Record<string, unknown>
> = DefaultEffectAPI<StoreState>[typeof SYMBOL_EFFECT_API_ROUTES];

/** @internal */
export function createEffectAPIDefaults<
  StoreState extends Record<string, unknown> = Record<string, unknown>
>({
  actions$,
  dispatch,
  getState,
  lifecycleRef,
  routesCollectionRef,
}: {
  actions$?: ActionSubject;
  dispatch: ReduxToolkit.Dispatch;
  getState: Redux.Store<StoreState>["getState"];
  lifecycleRef?: LifecycleRef<StoreState>;
  routesCollectionRef?: RoutesCollectionRef<StoreState>;
}): DefaultEffectAPI<StoreState> {
  return {
    actions$: actions$ || new ActionSubject(),
    dispatch,
    getState,
    [SYMBOL_EFFECT_API_LIFECYCLE]:
      lifecycleRef ||
      createMutableRef<MutableReferenceObjectValue<LifecycleRef>>(
        LIFECYCLE_REF_MESSAGE
      ),
    [SYMBOL_EFFECT_API_ROUTES]:
      routesCollectionRef ||
      createMutableRef<MutableReferenceObjectValue<RoutesCollectionRef>>(
        ROUTES_COLLECTION_REF_MESSAGE
      ),
  };
}

/**
 * A higher-order component that assigns props to the given component.
 *
 * @see {@link https://sables.dev/api#withprops withProps documentation}
 *
 * @example
 *
 * const FooLink = withProps("a", { className: "foo" });
 * const ButtonFooLink = withProps(LinkWithClassName, { role: "button" });
 *
 * @public
 */
export function withProps<C extends ElementType>(
  Component: C,
  baseProps: ComponentProps<C>
): FunctionComponent<ComponentProps<C>> {
  function ComponentWithProps(props: ComponentProps<C>) {
    return <Component {...baseProps} {...props} />;
  }

  const originalDisplayName =
    (typeof Component == "function" && Component.displayName) || "Component";

  ComponentWithProps.displayName = `withProps(${originalDisplayName})`;

  return ComponentWithProps;
}
