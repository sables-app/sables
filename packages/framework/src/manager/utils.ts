import {
  DefaultEffectAPI,
  SYMBOL_EFFECT_API_LIFECYCLE,
  SYMBOL_EFFECT_API_ROUTES,
  SYMBOL_MANAGER_EFFECT_API,
} from "@sables/core";
import { isSSREnv } from "@sables/utils";

import { LifecycleState, ServerRequestStateRef } from "../LifecycleContext.js";
import { ConfigureManagerFn, ConfigureManagerParams, Manager } from "../types.js";
import { createManager } from "./Manager.js";

export * from "@sables/utils";

function assertManagerConsistency<
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(manager: Manager<any, EffectAPI>) {
  const effectAPIRefOnManager = manager[SYMBOL_MANAGER_EFFECT_API];
  const lifecycleRefOnManager = manager[SYMBOL_EFFECT_API_LIFECYCLE];
  const routeCollectionRefOnManager = manager[SYMBOL_EFFECT_API_ROUTES];

  const effectAPI = effectAPIRefOnManager.demand();

  const lifecycleRefOnEffectAPI = effectAPI[SYMBOL_EFFECT_API_LIFECYCLE];
  const routeCollectionRefOnEffectAPI = effectAPI[SYMBOL_EFFECT_API_ROUTES];

  routeCollectionRefOnManager.demand();
  lifecycleRefOnManager.demand();
  lifecycleRefOnEffectAPI.demand();

  if (lifecycleRefOnManager.current !== lifecycleRefOnEffectAPI.current) {
    throw new Error("Inconsistent lifecycle state detected.");
  }

  if (
    routeCollectionRefOnEffectAPI.current &&
    routeCollectionRefOnEffectAPI.current !==
      routeCollectionRefOnManager.current
  ) {
    throw new Error(
      [
        "Multiple route collections detected.",
        "When a route collection object is defined on the effect API, it must match the one defined on the manager.",
      ].join("\n")
    );
  }
}

const DEFAULT_PARAMS: ConfigureManagerParams = {};

function getConfigureManagerParams(
  serverRequestStateRef: ServerRequestStateRef
): ConfigureManagerParams {
  if (!isSSREnv()) {
    return DEFAULT_PARAMS;
  }

  const { href } = serverRequestStateRef.demand();

  return {
    initialLocation: href,
  };
}

/** @internal */
export function resolveConfigureManager<M extends Manager<any, any>>(
  lifecycleState: LifecycleState,
  configureManager: ConfigureManagerFn<M> | undefined
): M {
  const { managerRef, serverRequestStateRef } = lifecycleState;

  if (!managerRef.current) {
    const managerCreator = configureManager || createManager;
    managerRef.current = managerCreator(
      getConfigureManagerParams(serverRequestStateRef)
    );
    managerRef.current[SYMBOL_EFFECT_API_LIFECYCLE].current = lifecycleState;
    assertManagerConsistency(managerRef.current);
  }

  return managerRef.current as M;
}

/**
 * A utility function to assist in defining `configureManager` functions.
 *
 * @example
 *
 * export const configureManager = defineConfigureManager(
 *   ({ initialLocation }) => {
 *     return createManager({ initialLocation });
 *   }
 * );
 *
 * @public
 */
export function defineConfigureManager<M extends Manager<any, any>>(
  configureManager: ConfigureManagerFn<M>
) {
  return configureManager;
}
