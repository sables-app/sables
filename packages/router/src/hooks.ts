import { useEffectAPI } from "@sables/core";

import { MouseEventHandler, useCallback, useMemo, useRef } from "react";
import { useDispatch } from "react-redux";

import { useRoutesCollection } from "./StoreEnhancer.js";
import type { BuildHrefParams, DynamicImportFn, RouteLink } from "./types.js";
import { createDynamicImportRegistrar } from "./utils.js";

/**
 * A React hook that returns an object for integrating a route with components.
 *
 * @example
 *
 * const routes = createRoutes()
 *   .set("root", "/")
 *   .set("profile", "/profile/:handle");
 *
 * function MyComponent() {
 *   const { ensureLocation } = useLink(routes.Profile, {
 *     handle: "@ash.ketchum",
 *   });
 *
 *   return (
 *     <button type="button" onClick={ensureLocation}>
 *       This link will cause the page to reload.
 *     </button>
 *   );
 * }
 *
 * @see {RouteLink}
 *
 * @public
 */
export function useLink(...buildHrefArgs: BuildHrefParams): RouteLink {
  const [id, params, opts] = buildHrefArgs;
  const dispatch = useDispatch();
  const routesCollection = useRoutesCollection();

  return useMemo(
    () => routesCollection.buildLink(dispatch, [id, params, opts]),
    [id, params, opts, dispatch, routesCollection]
  );
}

/**
 * A React hook that returns props for integrating a route with components
 * that render an anchor element.
 *
 * @example
 *
 * const routes = createRoutes()
 *   .set("root", "/")
 *   .set("profile", "/profile/:handle");
 *
 * function MyComponent(props: { onClick?: MouseEventHandler }) {
 *   const { props } = useLinkProps(props.onClick, routes.Profile, {
 *     handle: "@ash.ketchum"
 *   });
 *
 *   return <a {...props}>This link will cause a route transition.</a>;
 * }
 *
 * @see {RouteLink}
 *
 * @public
 */
export function useLinkProps(
  onClick?: MouseEventHandler<HTMLAnchorElement>,
  ...args: BuildHrefParams
) {
  const { href, ensureLocation } = useLink(...args);
  const handleClick = useCallback<MouseEventHandler<HTMLAnchorElement>>(
    (event) => {
      onClick?.(event);

      if (
        // It this behavior was prevented
        event.defaultPrevented ||
        // If the end-user is opening a new tab
        event.ctrlKey ||
        // If the end-user is opening a new window
        event.shiftKey ||
        // If a mouse button other than the
        // main left mouse button was clicked.
        event.button !== 0
      ) {
        // Do nothing, and let the default browser behavior do its thing
        return;
      }

      // Prevent browser from refreshing the window
      event.preventDefault();
      ensureLocation();
    },
    [ensureLocation, onClick]
  );

  return useMemo(
    () => ({
      handleClick,
      href,
      props: {
        href: href,
        onClick: handleClick,
      },
    }),
    [href, handleClick]
  );
}

/** @internal */
export function useRegisterDynamicImport(importer: DynamicImportFn) {
  const effectAPI = useEffectAPI();
  const isRegisteredRef = useRef(false);

  if (!isRegisteredRef.current) {
    isRegisteredRef.current = true;
    createDynamicImportRegistrar(effectAPI)(importer);
  }
}
