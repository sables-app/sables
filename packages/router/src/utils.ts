import {
  DefaultEffectAPI,
  SYMBOL_EFFECT_API_LIFECYCLE,
  SYMBOL_EFFECT_API_ROUTES,
} from "@sables/core";
import { isSSREnv } from "@sables/utils";

import type * as History from "history";
import { createPath } from "history";

import { ensureLocation, pushLocation, replaceLocation } from "./actions.js";
import { AnyRouteReference } from "./Routes.js";
import { isRoutesCollection, RoutesCollection } from "./RoutesCollection.js";
import type {
  BuildHrefParams,
  BuildLinkParams,
  LocationChangeAction,
  RegisterDynamicImportFn,
  RouteHref,
  RouteLink,
} from "./types.js";

/**
 * Used to compare location changes to determine whether
 * a route transition should start.
 * Only the pathname and search are compared. Hashes are not
 * considered to be a part of a route path.
 * @internal
 */
export function areLocationChangesRouterEquivalent(
  changeA?: LocationChangeAction["payload"],
  changeB?: LocationChangeAction["payload"]
): boolean {
  return !!(
    changeA &&
    changeB &&
    changeA.location.pathname === changeB.location.pathname &&
    changeA.location.search === changeB.location.search
  );
}

/** @internal */
export function isPartialHistoryPath(
  value?: AnyRouteReference | Partial<History.Path> | BuildHrefParams
): value is Partial<History.Path> {
  return (
    typeof value === "object" &&
    !Array.isArray(value) &&
    typeof (value as any).id === "undefined"
  );
}

/**
 * `import('history').createPath()` also returns the root
 * path when the given values are undefined.
 */
const FALLBACK_HREF = "/";

/**
 * Builds a hyperlink for a given route.
 *
 * @public
 */
export function buildHref(...args: BuildHrefParams): RouteHref {
  const [route, params, opts] = args;

  if (isPartialHistoryPath(route)) {
    return createPath(route) as RouteHref;
  }

  if (!route) return FALLBACK_HREF;

  return route.build(params, opts);
}

/**
 * Builds an object to interact with a route.
 *
 * @see {RouteLink}
 *
 * @public
 */
export function buildLink(...args: BuildLinkParams): RouteLink {
  const [dispatch, buildHrefArgs] = args;
  const href = buildHref(...buildHrefArgs);

  return {
    href,
    ensureLocation() {
      dispatch(ensureLocation(href));
    },
    pushLocation() {
      dispatch(pushLocation(href));
    },
    replaceLocation() {
      dispatch(replaceLocation(href));
    },
  };
}

/** @internal */
export function demandRoutesCollectionFromEffectAPI<
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(effectAPI: EffectAPI): RoutesCollection<EffectAPI> {
  const routesCollectionRef = effectAPI[SYMBOL_EFFECT_API_ROUTES];

  if (!isRoutesCollection<EffectAPI>(routesCollectionRef.current)) {
    throw new Error("Routes must be made available to use this effect.");
  }

  return routesCollectionRef.current;
}

/** @internal */
export function createDynamicImportRegistrar<
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(effectAPI: EffectAPI): RegisterDynamicImportFn {
  return function registerDynamicImport(dynamicImport) {
    if (!isSSREnv()) return;

    effectAPI[SYMBOL_EFFECT_API_LIFECYCLE].demand()
      .serverRequestStateRef.demand()
      .invokedImporters.add(String(dynamicImport));
  };
}
