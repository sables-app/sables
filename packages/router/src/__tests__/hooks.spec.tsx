import * as vitest from "vitest";
import { describe, expect, it } from "vitest";

import { useLink, useLinkProps } from "../hooks.js";
import { createRouterHookTester } from "./utils.js";

describe("hooks", () => {
  describe("useLink", () => {
    it("returns a route link payload", () => {
      const { render, resultRef } = createRouterHookTester(vitest, () =>
        useLink()
      );

      render();

      expect(resultRef.current).toMatchSnapshot();
    });
  });

  describe("useLinkProps", () => {
    it("returns a link props payload", () => {
      const { render, resultRef } = createRouterHookTester(vitest, () =>
        useLinkProps()
      );

      render();

      expect(resultRef.current).toMatchSnapshot();
    });
  });
});
