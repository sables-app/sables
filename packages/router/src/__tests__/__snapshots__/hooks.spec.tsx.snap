// Vitest Snapshot v1

exports[`hooks > useLink > returns a route link payload 1`] = `
{
  "ensureLocation": [Function],
  "href": "/",
  "pushLocation": [Function],
  "replaceLocation": [Function],
}
`;

exports[`hooks > useLinkProps > returns a link props payload 1`] = `
{
  "handleClick": [Function],
  "href": "/",
  "props": {
    "href": "/",
    "onClick": [Function],
  },
}
`;
